diff --git a/app/build.sh b/app/build.sh
index de092bd..cdce81b 100755
--- a/app/build.sh
+++ b/app/build.sh
@@ -296,7 +296,7 @@ elif [[ $BUILD_LINUX == 1 ]]; then
 	if [[ -n $arch ]]; then
 		omni_arch=$(get_canonical_arch l $arch)
 	else
-		for cand in x86_64 arm64 i686; do
+		for cand in arm64 x86_64 i686; do
 			[[ -d "${LINUX_RUNTIME_PATH_PREFIX}${cand}" ]] && { omni_arch="$cand"; break; }
 		done
 	fi
@@ -1004,7 +1004,7 @@ if [ $BUILD_LINUX == 1 ]; then
 	if [[ -n $arch ]]; then
 		archs=("$(get_canonical_arch l $arch)")
 	else
-		archs=(x64 arm64 i686)
+		archs=(arm64)
 	fi
 	for arch in "${archs[@]}"; do
 		[[ $arch == x64 ]] && arch="x86_64"
@@ -1041,6 +1038,7 @@ if [ $BUILD_LINUX == 1 ]; then
 		cp -RH "$CALLDIR/modules/zotero-libreoffice-integration/install" "$APPDIR/integration/libreoffice"

 		# Copy icons
+		mkdir -p "$APPDIR/icons/"
 		cp "$CALLDIR/linux/icons/icon32.png" "$APPDIR/icons/"
 		cp "$CALLDIR/linux/icons/icon64.png" "$APPDIR/icons/"
 		cp "$CALLDIR/linux/icons/icon128.png" "$APPDIR/icons/"
@@ -1052,7 +1050,7 @@ if [ $BUILD_LINUX == 1 ]; then

 		# Copy over removed-files and make a precomplete file
 		pushd "$APPDIR"
-		cp "$CALLDIR/update-packaging/removed-files_linux-$arch" removed-files
+		cp "$CALLDIR/update-packaging/removed-files_linux-x86_64" removed-files
 		python3 "$CALLDIR/scripts/createprecomplete.py"
 		if [ ! -s precomplete ]; then
 			echo "precomplete file not created -- aborting" 2>&1
diff --git a/app/config.sh b/app/config.sh
index 4a95b76..c4aca79 100644
--- a/app/config.sh
+++ b/app/config.sh
@@ -2,7 +2,7 @@ DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

 # Version of Gecko to build with
 GECKO_VERSION_MAC="140.6.0esr"
-GECKO_VERSION_LINUX="140.6.0esr"
+GECKO_VERSION_LINUX="140.6.0-1.el9"
 GECKO_VERSION_WIN="140.6.0esr"
 RUST_VERSION=1.86.0

diff --git a/app/scripts/fetch_xulrunner b/app/scripts/fetch_xulrunner
index 30c3a90..800c5a5 100755
--- a/app/scripts/fetch_xulrunner
+++ b/app/scripts/fetch_xulrunner
@@ -148,15 +148,6 @@ function modify_omni {
 	# Continue using app.update.auto in prefs.js on Windows
 	replace_line 'PER_INSTALLATION_PREFS_PLATFORMS = \["win"\]' 'PER_INSTALLATION_PREFS_PLATFORMS = []' modules/UpdateUtils.sys.mjs

-	# Prompt if major update is available instead of installing automatically on restart
-	replace_line 'if \(!updateAuto\) \{' 'if (update.type == "major") {
-      LOG("UpdateService:_selectAndInstallUpdate - prompting because it is a major update");
-      AUSTLMY.pingCheckCode(this._pingSuffix, AUSTLMY.CHK_SHOWPROMPT_PREF);
-      Services.obs.notifyObservers(update, "update-available", "show-prompt");
-      return;
-    }
-    if (!updateAuto) {' modules/UpdateService.sys.mjs
-
 	# Avoid console warning about resource://gre/modules/FxAccountsCommon.js
 	replace_line 'const logins = this._data.logins;' 'const logins = this._data.logins; if (this._data.logins.length != -1) return;' modules/LoginStore.sys.mjs

@@ -715,37 +706,54 @@ if [ $BUILD_LINUX == 1 ]; then
 	if [[ -n $arch ]]; then
 		arch_list=("$arch")
 	else
-		arch_list=(x86_64 arm64 i686)
+		arch_list=(arm64)
 	fi
 	for arch in "${arch_list[@]}"; do
 		xdir="firefox-linux-$arch"
 		rm -rf $xdir

-		archived_file="firefox-$GECKO_VERSION-$arch.tar.xz"
-		if [ -e "$archived_file" ]; then
-			echo "Using $archived_file"
-			cp "$archived_file" "firefox-$GECKO_VERSION.tar.xz"
-		else
-			if [ $arch = "arm64" ]; then
-				moz_arch=aarch64
+		if [ $arch = "arm64" ]; then
+			archived_file="firefox-$GECKO_VERSION.aarch64.rpm"
+			if [ -e "$archived_file" ]; then
+				echo "Using $archived_file"
+				cp "$archived_file" "firefox-$GECKO_VERSION.rpm"
 			else
-				moz_arch=$arch
+				# found a aarch64 build with version close to the x86_64 one, the centos version is static linked to icu, avoids version mismatch
+				curl -O "https://rpmfind.net/linux/centos-stream/9-stream/AppStream/aarch64/os/Packages/firefox-$GECKO_VERSION.aarch64.rpm"
+				cp "$archived_file" "firefox-$GECKO_VERSION.rpm"
 			fi
-			curl -O "$DOWNLOAD_URL/linux-$moz_arch/en-US/firefox-$GECKO_VERSION.tar.xz"
-			# Store local copy of unmodified app
-			cp "firefox-$GECKO_VERSION.tar.xz" "$archived_file"
+
+			mkdir -p tmp
+			bsdtar xvf firefox-$GECKO_VERSION.rpm -C tmp
+			mv tmp/usr/lib64/firefox $xdir
+			rm -r tmp
+			pushd $xdir
+			modify_omni $arch
+			popd
+			echo "$( $SCRIPT_DIR/xulrunner_hash -p l -a "$arch" )" > "hash-linux-${arch}"
+			rm "firefox-$GECKO_VERSION.rpm"
+		else
+			archived_file="firefox-$GECKO_VERSION-$arch.tar.xz"
+			if [ -e "$archived_file" ]; then
+				echo "Using $archived_file"
+				cp "$archived_file" "firefox-$GECKO_VERSION.tar.xz"
+			else
+				curl -O "$DOWNLOAD_URL/linux-$arch/en-US/firefox-$GECKO_VERSION.tar.xz"
+				# Store local copy of unmodified app
+				cp "firefox-$GECKO_VERSION.tar.xz" "$archived_file"
+			fi
+
+			tar xvf firefox-$GECKO_VERSION.tar.xz
+			mv firefox $xdir
+
+			pushd $xdir
+			modify_omni $arch
+			popd
+
+			rm "firefox-$GECKO_VERSION.tar.xz"
+
+			echo "$( $SCRIPT_DIR/xulrunner_hash -p l -a "$arch" )" > "hash-linux-${arch}"
 		fi
-
-		tar xvf firefox-$GECKO_VERSION.tar.xz
-		mv firefox $xdir
-
-		pushd $xdir
-		modify_omni $arch
-		popd
-
-		rm "firefox-$GECKO_VERSION.tar.xz"
-
-		echo "$( $SCRIPT_DIR/xulrunner_hash -p l -a "$arch" )" > "hash-linux-${arch}"
 	done
 fi
